crates/red_knot_python_semantic/src/types.rs

```diff
@@ -245,46 +263,77 @@ fn definition_expression_ty<'db>(
 fn bindings_ty<'db>(
     db: &'db dyn Db,
     bindings_with_constraints: BindingWithConstraintsIterator<'_, 'db>,
-) -> Option<Type<'db>> {
-    let mut def_types = bindings_with_constraints.map(
-        |BindingWithConstraints {
-             binding,
-             constraints,
-         }| {
-            let mut constraint_tys = constraints
-                .filter_map(|constraint| narrowing_constraint(db, constraint, binding))
-                .peekable();
-
-            let binding_ty = binding_ty(db, binding);
-            if constraint_tys.peek().is_some() {
-                constraint_tys
-                    .fold(
-                        IntersectionBuilder::new(db).add_positive(binding_ty),
-                        IntersectionBuilder::add_positive,
-                    )
-                    .build()
-            } else {
-                binding_ty
+) -> Symbol<'db> {
+    let mut is_unbound_visible = false;
+    let mut types = vec![];
```

Comment:
It seems like we shouldn't have to collect these into a vec? Can we make it a lazy map all the way through to the final union?

If not, then using a smallvec here might win back noticeable perf, since usually there should be few visible bindings.

Reply:
It improved performance by a tiny bit (<1%). And it's also a bit more readable, thanks.